File: onyx-web-database/src/main/java/com/onyx/server/DatabaseIdentityManager.java
package com.onyx.server;

import com.onyx.entity.SystemUser;
import com.onyx.exception.OnyxException;
import com.onyx.persistence.manager.PersistenceManager;
import io.undertow.security.idm.Account;
import io.undertow.security.idm.Credential;
import io.undertow.security.idm.IdentityManager;
import io.undertow.security.idm.PasswordCredential;

/**
 * This class is responsible for performing the account validation for http basic auth.
 *
 * This is an undertow implementation
 */
public class DatabaseIdentityManager implements IdentityManager {

    private final PersistenceManager persistenceManager;

    public DatabaseIdentityManager(PersistenceManager persistenceManager)
    {
        this.persistenceManager = persistenceManager;
    }

    @Override
    public Account verify(Account account) {
        // An existing account so for testing assume still valid.
        return account;
    }

    @Override
    public Account verify(String id, Credential credential) {
        Account account = getAccount(id);
        if (account != null && verifyCredential(account, credential)) {
            return account;
        }

        return null;
    }

    @Override
    public Account verify(Credential credential) {
        // TODO Auto-generated method stub
        return null;
    }

    private boolean verifyCredential(Account account, Credential credential)
    {
        if (credential instanceof PasswordCredential) {
            char[] password = ((PasswordCredential) credential).getPassword();
            final String sentPassword = new String(password);

            return account != null
                    && (sentPassword.equals(((DatabaseUserAccount) account).getPassword()));

        }
        return false;
    }

    private Account getAccount(final String id)
    {
        try
        {
            final SystemUser user = persistenceManager.findById(SystemUser.class, id);

            if(user == null)
                return null;

            return new DatabaseUserAccount(user);
        }
        catch (OnyxException e)
        {
            return null;
        }
    }

}

File: onyx-web-database/src/main/kotlin/com/onyx/server/DatabaseIdentityManager.kt
package com.onyx.server

import com.onyx.entity.SystemUser
import com.onyx.exception.OnyxException
import com.onyx.persistence.manager.PersistenceManager
import io.undertow.security.idm.Account
import io.undertow.security.idm.Credential
import io.undertow.security.idm.IdentityManager
import io.undertow.security.idm.PasswordCredential

/**
 * This class is responsible for performing the account validation for http basic auth.
 *
 * This is an undertow implementation
 */
class DatabaseIdentityManager(private val persistenceManager: PersistenceManager) : IdentityManager {

    override fun verify(account: Account): Account = account
    override fun verify(credential: Credential): Account? = null

    override fun verify(id: String, credential: Credential): Account? {
        val account = getAccount(id)
        return if (account != null && verifyCredential(account, credential)) {
            account
        } else null

    }

    private fun verifyCredential(account: Account?, credential: Credential): Boolean {
        if (credential is PasswordCredential) {
            val password = credential.password
            val sentPassword = String(password)

            return account != null && sentPassword == (account as DatabaseUserAccount).password

        }
        return false
    }

    private fun getAccount(id: String): Account? {
        return try {
            val user = persistenceManager.findById<SystemUser>(SystemUser::class.java, id) ?: return null
            DatabaseUserAccount(user)
        } catch (e: OnyxException) {
            null
        }
    }

}